src/components/DraggableLineChart/DraggableLineChartD3.ts

```diff
@@ -0,0 +1,186 @@
+			]);
+	}
+
+	drag = () => {
+		const { xScale, yScale } = this;
+		const { cx, onDragEnd } = this.params;
+		let initialPosition: number;
+		return d3Drag
+			.drag()
+			.on('start', function() {
+				const activeDot = d3Selection.select(this);
+				initialPosition = Number(activeDot.attr('cy'));
+			})
+			.on('drag', function(d: any) {
+				const [max, min] = yScale.range();
+				const activeDot = d3Selection.select(this);
+				const adjMouseY = initialPosition + d3Selection.event.y;
+				const newPointY =
+					adjMouseY < min ? min : adjMouseY > max ? max : adjMouseY;
+				const lines = d3Selection.selectAll(`.${cx('&-Line')}`);
+
+				d.y = Number(yScale.invert(newPointY));
+				activeDot.attr('cy', newPointY);
+				lines.attr(
+					'd',
+					// @ts-ignore
+					d3Shape
+						.line()
+						.x(function({ x }: any) {
+							return xScale(x);
+						})
+						.y(function({ y }: any) {
+							return yScale(y);
+						})
+				);
+			})
+			.on('end', onDragEnd || _.noop);
+	};
+	renderXAxis = () => {
+		this.selection.append('g').call((xAxis: any) => {
+			const { margin, height, cx } = this.params;
+			xAxis
+				.attr('transform', `translate(${0},${margin.top})`)
+				.classed(`${cx('&-Axis')}`, true)
+				.classed('xAxis', true)
+				.call(
+					d3Axis
+						.axisTop(this.xScale)
+						.tickSize(margin.top + margin.bottom - height)
+				);
+		});
+	};
+	renderYAxis = () => {
+		let yAxisGroup = this.selection.selectAll('.yAxisGroup');
+		if (yAxisGroup.empty()) {
+			yAxisGroup = this.selection.append('g').classed('yAxisGroup', true);
+		}
+		yAxisGroup.call((yAxis: any) => {
+			const { margin, cx } = this.params;
+			yAxis
+				.attr('transform', `translate(${margin.left},${0})`)
+				.classed(`${cx('&-Axis')}`, true)
+				.transition()
+				.duration(500)
+				.call(d3Axis.axisLeft(this.yScale).ticks(10));
+		});
+	};
+	renderLine = (isNew: boolean) => {
+		const { cx } = this.params;
+		const lines = isNew
+			? this.selection
+					.append('g')
+					.append('path')
+					.attr('class', `${cx('&-Line')}`)
+			: this.selection.selectAll(`.${cx('&-Line')}`);
+		lines.datum(this.params.data).enter();
+		lines
+			.transition(d3Transition.transition().duration(500))
+			.attr('fill', 'none')
+			.attr(
+				'd',
+				d3Shape
+					.line()
+					.x((d: any) => this.xScale(d.x))
+					.y((d: any) => this.yScale(d.y))
+			);
+	};
+	renderPoints = (isNew: boolean) => {
+		const { data } = this.params;
+		const circle = isNew
+			? this.selection
+					.append('g')
+					.selectAll('circle')
+					.data(data)
+					.join('circle')
+			: this.selection
+					.selectAll('circle')
+					.data(data)
+					.join('circle');
+
+		circle
+			.transition()
+			.duration(500)
+			.attr('cx', (d: any) => this.xScale(d.x))
+			.attr('cy', (d: any) => this.yScale(d.y))
+			.attr('r', 5)
+			.style('fill', '#009fdb')
+			.style('stroke', 'white')
+			.style('stroke-width', 1);
+		if (isNew) circle.call(this.drag());
+	};
+	renderLineChart = () => {
```

Comment:
any reason we don't separate all of these 'renders' to their own react component? Was it issues with lifecycles? It feels a little code by config, which is hard to maintain.